In [54]:
import pandas as pd
import plotly.express as px

car_data = pd.read_csv('../vehicles.csv') # lendo os dados

# ANÁLISE

In [55]:
car_data.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [56]:
car_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [57]:
car_data.head()

price  model_year           model  condition  cylinders fuel  odometer  \
0   9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1  25500         NaN      ford f-150       good        6.0  gas   88705.0   
2   5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3   1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4  14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   

  transmission    type paint_color  is_4wd date_posted  days_listed  
0    automatic     SUV         NaN     1.0  2018-06-23           19  
1    automatic  pickup       white     1.0  2018-10-19           50  
2    automatic   sedan         red     NaN  2019-02-07           79  
3    automatic  pickup         NaN     NaN  2019-03-22            9  
4    automatic   sedan       black     NaN  2019-04-02           28

In [58]:
car_data.duplicated().sum()

0

In [59]:
car_data.corr(numeric_only=True) # parametro para apenas aceitar colunas numéricas

price  model_year  cylinders  odometer  is_4wd  days_listed
price        1.000000    0.428656   0.296627 -0.420299     NaN    -0.000682
model_year   0.428656    1.000000  -0.148040 -0.473948     NaN    -0.005703
cylinders    0.296627   -0.148040   1.000000  0.090160     NaN     0.005517
odometer    -0.420299   -0.473948   0.090160  1.000000     NaN     0.003804
is_4wd            NaN         NaN        NaN       NaN     NaN          NaN
days_listed -0.000682   -0.005703   0.005517  0.003804     NaN     1.000000

In [60]:
car_data.isna().sum()

price               0
model_year       3619
model               0
condition           0
cylinders        5260
fuel                0
odometer         7892
transmission        0
type                0
paint_color      9267
is_4wd          25953
date_posted         0
days_listed         0
dtype: int64

# VALORES AUSENTES
1. Trocar os valores ausentes em **is_4wd** para 0 (ou seja, é uma coluna booleana sendo 1 = 4x4 e 0 = tração em 2 rodas)

2. Preencher os valores ausentes em **model_year** de acordo com odômetro e preço

3. Preencher os valores ausentes em **odometer** de acordo com o ano do modelo e condição

4. Preencher os valores ausentes em **cylinders** 

5. Preencher os valores ausentes em **paint_color** com o nome "desconhecido"

In [61]:
car_data['is_4wd'] = car_data['is_4wd'].fillna(0.0)

In [62]:
car_data.loc[car_data['model_year'].isna()].head()

price  model_year            model  condition  cylinders fuel  odometer  \
1   25500         NaN       ford f-150       good        6.0  gas   88705.0   
20   6990         NaN  chevrolet tahoe  excellent        8.0  gas  147485.0   
65  12800         NaN       ford f-150  excellent        6.0  gas  108500.0   
69   7800         NaN       ford f-150   like new        8.0  gas   97510.0   
72   3650         NaN   subaru impreza  excellent        NaN  gas   74000.0   

   transmission    type paint_color  is_4wd date_posted  days_listed  
1     automatic  pickup       white     1.0  2018-10-19           50  
20    automatic     SUV      silver     1.0  2018-08-05           28  
65    automatic  pickup       white     0.0  2018-09-23           15  
69    automatic   truck       white     1.0  2019-02-20           39  
72    automatic   sedan        blue     1.0  2018-08-07           60

In [63]:
car_data.loc[car_data['odometer'] >= 200000.0].median(numeric_only=True) 
# a mediana de do ano de carros que tem mais de 200k de km rodados é 2004

price            4500.0
model_year       2004.0
cylinders           8.0
odometer       226998.0
is_4wd              1.0
days_listed        33.0
dtype: float64

In [64]:
car_data.loc[
    (car_data['odometer'] > 200000.0) & 
    (car_data['model_year'].isna()) , 'model_year'] = 2004

In [65]:
# repetindo o mesmo processo
car_data.loc[car_data['odometer'] <= 40000].median(numeric_only=True) 

price          19995.0
model_year      2017.0
cylinders          6.0
odometer       21582.0
is_4wd             0.0
days_listed       33.0
dtype: float64

In [66]:
car_data.loc[
    (car_data['odometer'] <= 40000.0) & 
    (car_data['model_year'].isna()) , 'model_year'] = 2017.0

In [67]:
car_data.loc[(car_data['odometer'] >= 40000.0) & (car_data['odometer'] <= 100000.0)].median(numeric_only=True) 

price          13597.0
model_year      2014.0
cylinders          6.0
odometer       74520.5
is_4wd             0.0
days_listed       33.0
dtype: float64

In [68]:
car_data.loc[
    (car_data['odometer'] >= 40000.0) & (car_data['odometer'] <= 100000.0) & 
    (car_data['model_year'].isna()), 'model_year'] = 2014.0

In [69]:
car_data.loc[(car_data['odometer'] >= 100000) & (car_data['odometer'] <= 200000)].median(numeric_only=True) 

price            6995.0
model_year       2009.0
cylinders           6.0
odometer       139250.0
is_4wd              1.0
days_listed        33.0
dtype: float64

In [70]:
car_data.loc[
    (car_data['odometer'] >= 100000) & (car_data['odometer'] <= 200000) & 
    (car_data['model_year'].isna()), 'model_year'] = 2009.0

In [71]:
car_data.groupby('model_year')['price'].median().tail(20)
# de 2000 a 2005 os valores são abaixo de 5000

model_year
2000.0     3500.0
2001.0     3995.0
2002.0     3999.0
2003.0     4488.0
2004.0     4999.0
2005.0     4995.0
2006.0     5900.0
2007.0     6050.0
2008.0     6995.0
2009.0     6950.0
2010.0     7995.0
2011.0     9995.0
2012.0     9995.0
2013.0    11995.0
2014.0    13990.0
2015.0    17700.0
2016.0    18000.0
2017.0    19700.0
2018.0    23800.0
2019.0    23800.0
Name: price, dtype: float64

In [72]:
car_data.loc[
    (car_data['price'] < 5000) & 
    (car_data['model_year'].isna()), 'model_year'] = 2002.0

In [73]:
car_data.loc[
    (car_data['price'] > 10000) & (car_data['price'] < 20000) &
    (car_data['model_year'].isna()), 'model_year'] = 2015.0

In [74]:
car_data.loc[
    (car_data['price'] > 20000) &
    (car_data['model_year'].isna()), 'model_year'] = 2019.0

In [75]:
# começar a preencher a coluna 'odometer'
car_data.groupby('odometer')['model_year'].median()

odometer
0.0         2010.0
1.0         2007.0
2.0         2005.0
3.0         2011.0
5.0         2018.0
             ...  
840000.0    2010.0
866000.0    2014.0
920000.0    1996.0
980000.0    2014.0
990000.0    1988.5
Name: model_year, Length: 17762, dtype: float64

In [76]:
# filtrando os odômetros de 1 a 1000 km e vendo a média do ano de fabricação com base neles
um_mil_km = car_data[(car_data['odometer'] > 0) & (car_data['odometer'] <= 1000)]

media_ano = um_mil_km['model_year'].mean()
media_ano

2008.6605783866057

In [77]:
# a mediana dos odômetros dos carros 'new' é 8000
car_data[(~car_data['odometer'].isna()) & (car_data['condition'] == 'new')].median(numeric_only=True)


price          22525.0
model_year      2018.0
cylinders          6.0
odometer        8002.0
is_4wd             0.0
days_listed       35.0
dtype: float64

In [78]:
car_data.loc[
    (car_data['condition'] == 'new') &
    (car_data['odometer'].isna()), 'odometer'] = 8000.0

In [79]:
# preenchendo os valores ausentes com o método "tranform" com base na mediana do odômetro de acordo com o modelo e a condição
car_data['odometer'] = car_data['odometer'].fillna(
    car_data.groupby(['condition', 'model_year'])['odometer'].transform('median')
)

In [80]:
# começando a coluna cylinders
car_data[(car_data['model'] == 'honda pilot') & ~car_data['cylinders'].isna()]


price  model_year        model  condition  cylinders fuel  odometer  \
7      15990      2013.0  honda pilot  excellent        6.0  gas  109473.0   
93     32000      2017.0  honda pilot   like new        6.0  gas   32104.0   
343     4000      2005.0  honda pilot  excellent        6.0  gas  167000.0   
598    13000      2011.0  honda pilot  excellent        6.0  gas  123048.0   
655     1999      2007.0  honda pilot       good        6.0  gas  288000.0   
...      ...         ...          ...        ...        ...  ...       ...   
50448   4000      2003.0  honda pilot       good        4.0  gas  220000.0   
50487   9995      2013.0  honda pilot       good        6.0  gas  127914.0   
50708   8346      2009.0  honda pilot  excellent        6.0  gas  130000.0   
50919   4900      2009.0  honda pilot       good        6.0  gas  138348.0   
51416   5490      2005.0  honda pilot       good        6.0  gas  190333.0   

      transmission type paint_color  is_4wd date_posted  days_listed  
7        automatic  SUV       black     1.0  2019-01-07           68  
93       automatic  SUV       white     0.0  2018-07-12          100  
343      automatic  SUV         red     1.0  2018-05-12            8  
598      automatic  SUV        grey     1.0  2018-07-19            3  
655      automatic  SUV      custom     1.0  2018-12-18           83  
...            ...  ...         ...     ...         ...          ...  
50448    automatic  SUV       brown     1.0  2018-05-28           22  
50487    automatic  SUV        grey     0.0  2018-07-10           30  
50708    automatic  SUV        blue     1.0  2018-05-03           29  
50919    automatic  SUV      silver     1.0  2018-06-10            7  
51416    automatic  SUV      silver     1.0  2018-10-14           93  

[274 rows x 13 columns]

In [81]:
car_data.loc[car_data['model'] == 'honda pilot'].groupby('cylinders').agg('count')
# analisando, descobri que cada modelo segue uma ordem de cilindros, 
# ou seja, se a moda de cilidros de um modelo for X, é provavel que um carro do mesmo modelo tenha X cilindros

price  model_year  model  condition  fuel  odometer  transmission  \
cylinders                                                                      
4.0            2           2      2          2     2         2             2   
5.0            1           1      1          1     1         1             1   
6.0          269         269    269        269   269       269           269   
8.0            2           2      2          2     2         2             2   

           type  paint_color  is_4wd  date_posted  days_listed  
cylinders                                                       
4.0           2            2       2            2            2  
5.0           1            1       1            1            1  
6.0         269          225     269          269          269  
8.0           2            2       2            2            2

In [82]:
# fazendo com que os valores ausentes de cilindros sejam substituídos pela moda do modelo
# se não houver moda, usa-se a mediana de todos os modelos
car_data['cylinders'] = car_data['cylinders'].fillna(
    car_data.groupby('model')['cylinders'].transform(lambda x: x.mode()[0] if not x.mode().empty else x.median())
)


In [83]:
# preenchendo a coluna paint_color
car_data['paint_color'] = car_data['paint_color'].fillna('unknown')

In [84]:
# por fim, excluir as poucas linhas faltantes
car_data = car_data.dropna()

In [85]:
car_data.isna().sum()

price           0
model_year      0
model           0
condition       0
cylinders       0
fuel            0
odometer        0
transmission    0
type            0
paint_color     0
is_4wd          0
date_posted     0
days_listed     0
dtype: int64

# TIPOS DE COLUNAS
1. Trocar a coluna **model_year** de float para **int**

2. Trocar a coluna **cylinders** de float para **int**

3. Trocar a coluna **odometer** de float para **int**

4. Trocar a coluna **is_4wd** de float para **int**

5. Trocar a coluna **date_posted** de object para **datetime**  

In [86]:
# convertendo as colunas de float para int
car_data['model_year'] = car_data['model_year'].astype('int')
car_data['cylinders'] = car_data['cylinders'].astype('int')
car_data['odometer'] = car_data['odometer'].astype('int')
car_data['is_4wd'] = car_data['is_4wd'].astype('int')

# convertendo para datetime
car_data['date_posted'] = pd.to_datetime(car_data['date_posted'])

# ADICIONAR COLUNAS FILTRADAS

In [87]:
# criando uma coluna apenas para o ano em que foi vendido
car_data['year_posted'] = car_data['date_posted'].dt.year 

In [88]:
# criando coluna apenas para a marca de cada carro
car_data['brand'] = car_data['model'].str.split().str[0]

In [89]:
# criando a idade de um carro
car_data['car_age'] = car_data['year_posted'] - car_data['model_year']

In [90]:
# código que irá para a visualização do site
q1 = car_data['days_listed'].quantile(0.25)
q3 = car_data['days_listed'].quantile(0.75)
iqr = q3 - q1

outliers = car_data[(car_data['days_listed'] < (q1 - 1.5 * iqr)) | (car_data['days_listed'] > (q3 + 1.5 * iqr))]
outliers['brand'].value_counts() / car_data['brand'].value_counts() * 100

brand
acura            2.966102
bmw              4.905660
buick            2.611940
cadillac         4.049844
chevrolet        3.166352
chrysler         3.592814
dodge            3.120000
ford             3.096294
gmc              3.157895
honda            3.339090
hyundai          2.485004
jeep             3.241590
kia              3.436426
mercedes-benz         NaN
nissan           2.911709
ram              3.174123
subaru           3.238547
toyota           3.061036
volkswagen       3.352601
Name: count, dtype: float64

# TRATAMENTO DE ERRO
1. Ao fazer a EDA, percebi que 800 carros estavam com o preço de 1 dolar, isso pode acontecer por N motivos, erro de digitação, erro ao computar os dados ou até uma forma de doação. Então, para tratar esse erro e não atrapalhar os dados, resolvi tirar todos os carros abaixo de 150 dólares, apagando apenas 900 e deixando o Dataframe mais limpo.

2. Retirando a marca "mercedes-benz" pois tem apenas 41 dados, todos são do mesmo modelo, mesmo preço, cilindros, combustivel, transmissao, etc. As linhas não foram detectadas no "duplicated()" pois as datas variam.

In [91]:
car_data = car_data[car_data['price'] > 100]

In [92]:
car_data.loc[car_data['brand'] == 'mercedes-benz'].head(10)

price  model_year                              model  condition  \
42     34900        2013  mercedes-benz benze sprinter 2500  excellent   
1642   34900        2013  mercedes-benz benze sprinter 2500  excellent   
2232   34900        2013  mercedes-benz benze sprinter 2500  excellent   
2731   34900        2013  mercedes-benz benze sprinter 2500  excellent   
4149   34900        2019  mercedes-benz benze sprinter 2500  excellent   
4681   34900        2013  mercedes-benz benze sprinter 2500  excellent   
5681   34900        2013  mercedes-benz benze sprinter 2500  excellent   
8975   34900        2013  mercedes-benz benze sprinter 2500  excellent   
10600  34900        2013  mercedes-benz benze sprinter 2500  excellent   
11541  34900        2019  mercedes-benz benze sprinter 2500  excellent   

       cylinders    fuel  odometer transmission type paint_color  is_4wd  \
42             6  diesel     98000    automatic  van       black       0   
1642           6  diesel     98000    automatic  van       black       0   
2232           6  diesel     98000    automatic  van       black       0   
2731           6  diesel     98000    automatic  van       black       0   
4149           6  diesel     14531    automatic  van       black       0   
4681           6  diesel     98000    automatic  van     unknown       0   
5681           6  diesel     98000    automatic  van       black       0   
8975           6  diesel     98000    automatic  van     unknown       0   
10600          6  diesel     98000    automatic  van       black       0   
11541          6  diesel     14531    automatic  van     unknown       0   

      date_posted  days_listed  year_posted          brand  car_age  
42     2019-01-15           16         2019  mercedes-benz        6  
1642   2018-12-04           36         2018  mercedes-benz        5  
2232   2018-08-23           70         2018  mercedes-benz        5  
2731   2019-04-12           31         2019  mercedes-benz        6  
4149   2018-10-12           28         2018  mercedes-benz       -1  
4681   2018-10-02           32         2018  mercedes-benz        5  
5681   2018-12-11           34         2018  mercedes-benz        5  
8975   2018-09-24           45         2018  mercedes-benz        5  
10600  2018-09-16           47         2018  mercedes-benz        5  
11541  2018-05-28           24         2018  mercedes-benz       -1

In [93]:
car_data = car_data[car_data['brand'] != 'mercedes-benz']

# DATAFRAME LIMPO

In [94]:
car_data.to_csv('car_data.csv', index=False)